### ISSUE TITLE
Ore spawn improvements

### ISSUE THREAD

**Issue by Fixer-007** [minetest/minetest_game#1168]
I did some limited research on minetest ore spawning:
- all valuable ores can be mined easily at depths below -256
- rarest of all `MESE block` spawns below -1024 (but is not required, since crystals can be mined past -256 anyway and combined into MESE block)
- minetest has 30000 of depth of uniformly distributed ores below -256

Now that's big disappointment for miners, so... why not we add even more layers **below -1024**? Say at -4000, -16000, -32000 with increased probability of those ores.

@kwolekr mentioned that nobody pushed idea about biome specific ores, now that's great idea that adds more challenge for players, you can make it like:
- coal, iron - all biomes (if you are insane, go with coal in forest biomes)
- copper - taiga/tundra/glacier // Who uses copper anyway? :}
- gold, diamond, mese - desert, savannah (South Africa as a gold and diamond capital of the world).

That will seriously change mining gameplay, since you will need to actually travel/search for needed ores. Especially on server with `protector redo` used for protection. New players will be confused though.

biome map: https://cloud.githubusercontent.com/assets/3686677/15913099/f8b759ac-2dcf-11e6-84ff-ae61176010b1.png

**Comment by MarkuBu** [comment]
Not everybody wants to digg down thousands of nodes. Should be configurable

But I like the idea of biome specific ores. Had a similar idea

**Comment by Fixer-007** [comment]
>Not everybody wants to digg down thousands of nodes.

No, only addition is EVEN more ores at bigger depths, nothing else changes.

**Comment by C1ffisme** [comment]
I once had the idea of changing how ores work so that some exploration is needed.

There are three tiers of tools, armor and blocks. Each one is stronger than the last. There are also some in-between tiers for decorative or non-mining use.

We have 3 main ores, copper, iron and tin (or maybe zinc?).

There are 3 layers of ground, each having a different type of rock: Sedimentary, Metamorphic and Igneous, each one stronger than the last.

<h4>Tier 0, Stone Pickaxe:<h4>

A stone pickaxe can be constructed by lashing 2 stones and a stick together with rope. Rope is made from a fiber-like plant ([Sisal](https://en.wikipedia.org/wiki/Sisal#Uses), maybe?), stones can be found on the ground, and wood can be found by smashing as trees with a stone. (Or we could punch it, if you want to keep that feature.)

A stone pickaxe can mine tier 1 ores, and Sedimentary rock.

<h4>Tier 1, Singular Metals:<h4>

Pretty simple. Each of the three ores can be mined up from the first layer of rock, which is Sedimentary. Once you smelt these metals, you can build a tier 1 pickaxe out of them. Tier 1 pickaxes can mine Metamorphic stone and tier 2 ores.

<h4>Tier 2, First Alloys:<h4>

Using clay on the surface, you can build a crucible. This becomes useful later when you start mining in Metamorphic rocks.

Like I said, other ores may spawn here for other purposes, but the most important one is coal. Coal is a very pure form of carbon, and can be used for sticking metals together. You can put together two of any of the three metals, and get Tier 2 ingots, to make Tier 2 tools. It does not matter which ones you use, they will all work. (But you cannot stick all 3 together, not yet.)

A Tier 3 pickaxe can mine Igneous rock, and Tier 3 ores.

<h4>Tier 3, Steel Making:<h4>

Some old volcanic activity long ago has left quite a bit of Igneous rock at the bottom of the world.

Steel cannot just be made from coal and iron, it requires the other two metals, and one more thing.

In the Igneous rock, you will find Tungsten, which can be used to make Tier 4/Steel tools. This is the final tier (without mods). And they are the only tools that can mine Obsidian, which is also found in the Igneous layer. (Very useful for making near-invincible shelters.)

Steel is made by putting Tungsten, Iron, Coal, Tin/Zinc and Copper into the Crucible. However, a larger amount of Iron is needed than the other metals.

Might need some balancing/tweaking, but here is what the recipe for Steel would require:

- 1 Copper lump
- 1 Tin/Zinc lump
- 2 Coal lumps
- 2 Tungsten lumps
- 5 Iron lumps

<h4>How ores would spawn:<h4>

Copper, Iron and Tin/Zinc would not spawn in the traditional way. Each layer is divided into "underground biomes", there are three of these "biomes", with one of each of the ores in them.

Copper, Tin/Zinc all spawn in the Sedimentary and Metamorphic layers, but not in the Igneous layer. Iron spawns in all layers, because of it's importance in making steel.

Also, Iron "biomes" are less common in the Sedimentary layer, as common as Copper and Tin/Zinc biomes in the Metamorphic layer, and, of course, everywhere in the Igneous layer.

<h4>Pros and Cons of this system:<h4>

**Pros: **

- Forces players to build methods of transportation to spots with different metal deposits.
- Allows for more cooperation in multiplayer, since the game will go quicker if two people are searching for metals in different areas.
- PVP servers will get more interesting, since people will have to cooperate if they want to fight the one player with the Obsidian castle.

**Cons: **

- Not sure if we can change Minetest Game to this different game without doing some harm to servers, I'd rather this be another subgame.
- Some people might find it hard to get help on a server.
- We would need to implement a few more features to help players along, such as transportation venues or mining machines, otherwise the game would take a bit too long.

<h4>Other Ideas:<h4>

- Make each metal have certain uses, such as Iron being used for armor, Copper being used for mechanics, and Tin/Zinc for, I dunno, Animal/Mob control? (Perhaps other metals scare animals, and tin can be used for animal pens or something?)
- **AND** each alloy allows a combination of those uses, such as Iron + Copper (Gnucydur) = Defensive Machines, or Mechanically enhanced armor, Tin/Zinc + Copper (Bronze/Brass) = Automatic Animal care machines, Iron + Tin/Zinc (Pewter???) = Some kind of Animal Armor, like [Horse Armor](http://minecraft.gamepedia.com/Horse_Armor), maybe? PROBLEM : I intended for all Tier X ingots to be the same. Smelting together any Tier 1 metals would always make a Tier 2 ingot, not a Gnucydur ingot, or a Bronze/Brass ingot, or a Pewter Ingot.
- Gold could be found in the Metamorphic layer, and might allow some sort of magic/enchanting/enhancing.
- Diamonds in the Igneous layer could be used for Industrial Mining, you couldn't slap it on your pickaxe, but you can tip a mining drill with one. Perhaps this would be the only way to smash through a Tier 5 layer? (Of course, a Tier 4 layer would be mine-able by a Steel Pick, because of consistency.)

Thoughts, anyone?

**Comment by paramat** [comment]
^ That's a different subgame.

Note everything under y = -112 is a single biome called 'underground', so restricting ores to biomes only works above that, making those ores easy to find.
Currently all mapgens have identical ore spawning, the same as mgv6. Any changes have to decide which mapgens are affected.

**Comment by C1ffisme** [comment]
@paramat It's a better system than the current one, though. Mind you, we would have to add underground "biomes" and layers of rock too.

**Comment by NathanSalapat** [comment]
@Fixer-007 Copper is used in nearly every machine with the technic mod, so it is a very used and needed ore.

**Comment by 0-afflatus** [comment]
This is essentially what I do in grailtest. I use five different rock types, which different ores spawn in, roughly corresponding to Sedimentary, Metamorphic and Igneous types. Apart from desert stone, none are determined by biome. The first 1k is mixed. There is an expectation that mining should get more productive the deeper you go, but I think it adds to gameplay if there is a sense that different (underground) environments yield different minerals, so you can do more focused mining. Placement of resources hugely affects gameplay, so there is a limit to how far you can take this with MTG. Adding more layers from -1k to around -5k is the best way to go, so long as game balance is respected, naturally.

**Comment by paramat** [comment]
> MESE block spawns below -1024 (but is not required, since crystals can be mined past -256 anyway and combined into MESE block)

This makes me think we should not allow combining mese crystal into mese block, to make mese block meaningful and a higher gameplay cost if a recipe requires it.

> add even more layers below -1024? Say at -4000, -16000, -32000 with increased probability of those ores.

Could do, probably hasn't just to keep number of ore registrations low.
Remember that with future 'stacked realms' another realm may start a few thousand down, and anything much more than 1000 down is possibly too much of a challenge for some players, so perhaps we should not bother changing anything under y = -1000.

**Comment by Fixer-007** [comment]
>Could do, probably hasn't just to keep number of ore registrations low.
>Remember that with future 'stacked realms' another realm may start a few thousand down, and anything much more than 1000 down is possibly too much of a challenge for some players, so perhaps we should not bother changing anything under y = -1000.

Is not with stacked realms ore gen will be reworked anyway? Since there will be multiple starting points, also, as I see it... if realm has 2000 of above and say 500 of underground, ore will be generated for 0-500, if underground is larger, like 0-5000, it will generate ores up to 5000 adjusting automatically in some way.

**Comment by paramat** [comment]
Actually i mean right now we have lua mapgens creating underworlds that tend to start a few thousand down.

It might be good to spread the depths more progressively, so have something starting at y = -500.

**Comment by 0-afflatus** [comment]
> anything much more than 1000 down is possibly too much of a challenge for some players, so perhaps we should not bother changing anything under y = -1000.

Having watched players' mining activity I disagree with this POV very strongly. 

The majority of players can easily get 500 to 1k down and they should be rewarded with lots of coal, iron and less valuable ores. If you're collecting, say, diamonds you should have to go deeper to get any quantity - strong players will drill down to -5k in less than a week of play and they should be amply rewarded for their efforts. Some minerals (like coal) should run out or get scarce at deeper levels. This leads to different styles of mining for different elements, ideally coal mines should be shallow and chase seams, diamond mines should be deep and chase veins (and yes, I do know how difficult they are to get right). There are a lot of oregen options that MTG isn't using - and that's another reason for adding more registrations for -1k to -5k.

> we have lua mapgens creating underworlds that tend to start a few thousand down.

They need to be able to redefine oregen so they don't get clobbered by default settings. I think that's trivial to deal with.

**Comment by paramat** [comment]
Good points.

**Comment by paramat** [comment]
I have simplified ore registrations a little to reduce them to 3 per ore, so we can now afford to add 1 more increase in density at a lower depth, i like this idea, but only 1 more to keep it simple. I might work on this myself.

**Comment by paramat** [comment]
There is also an argument for making ores deeper, as most mapgens now have never-ending tunnels that can take you down to y = -256 with almost no effort. In mgv6 lots of digging was required.
So, perhaps this can be a good chance to change all the ore depths to be deeper and a more consistent progression? I'm happy to do this sometime. Then we could also add another increase at a lower depth when doing this.

**Comment by C1ffisme** [comment]
Little bit offtopic, but if we're going to move ores lower, wouldn't it make sense to fill the area in-between with something interesting to look at?

**Comment by paramat** [comment]
Hehe we have tunnels, large caves with lava and water, and massive caverns already.
We're talking about moving the 'depths at which ores become more dense' lower, there will not be gaps without ores anywhere, nothing will be removed.

**Comment by C1ffisme** [comment]
@paramat No, I don't think there will be gaps of stone without ores, but considering an ore is simply the stone texture, but with small specks of color in them, I think caves and such need a bit more decoration before we start forcing players to travel through them longer.

**Comment by paramat** [comment]
Decoration will come when we have underground decorations in the engine Biome API, which are not possible yet. Making underground more interesting is already planned.

**Comment by paramat** [comment]
My suggested plan:

* Make some ores a little deeper, especially valuable ones, to compensate for easier vertical travel made possible by 3D noise tunnels. This is also consistent with how surface terrain and biomes are now larger scale then mgv6.
* Simplify iron ore placement.  Currently it has 2 registrations that, unlike all other ores, actually overlap at lower y, we can combine these 2 into 1 with no change in amount of ore at lower y.
* Make ore depth progressive with value with this order:
Coal, Iron, Tin, Copper, Gold, Mese ore, Diamond, Mese block.
* Add a 3rd increase in ore density, as requested, at much lower depths.

**Comment by ThomasMonroe314** [comment]
I agree with the 3rd stage of density and the progressive depth

**Comment by paramat** [comment]
PR #1168

**Comment by DS-Minetest** [comment]
@paramat #1168 is this issue, do you mean #1810?

**Comment by paramat** [comment]
Thanks.

**Comment by paramat** [comment]
It's amazing how all valuable ores except mese block are found above y = -128, with the 3D noise tunnels it's possible to walk to that depth in a game session.
Our wider 3D noise tunnels also make ores much more common in tunnel walls so this is another way mining has been made too easy in non-mgv6 mapgens.
Another thing to consider is how mese ore can be combined into mese block, so the two should probably occur at the same or similar depths, mese is very high value like diamond so mese ore at y = -64 is ridiculous.

**Comment by DS-Minetest** [comment]
>mese ore can be combined into mese block, so the two should probably occur at the same or similar depths

Only in deep depths there's enough pressure and mese crystal concentration to be compressed into block form.

**Comment by paramat** [comment]
I see it the other way around, mese ore is shattered mese block, the letters could not form under pressure.

**Comment by paramat** [comment]
2nd PR #1813

**Comment by paramat** [comment]
#1813 merged.

Still to do, perhaps add a 2nd increase in density at a deeper level.

**Comment by paramat** [comment]
This is my job, i will add the extra density increase, it's easy. Not really suitable to throw out to contributors.

**Comment by paramat** [comment]
So now the lowest ore density increase is Mese blocks at y = -2048. What depth should the next final density increase be out of -4000, -8000 or -16000 (halfway to world base)? I feel that -4000 may be too easy and remove reward for going very deep, any opinions?

**Comment by Fixer-007** [comment]
It should probably be pretty challenging, like -16000/-20000/-24000/...

**Comment by paramat** [comment]
I'll rule out -4000.
Trying to decide between -8000 and -16000.

**Comment by ThomasMonroe314** [comment]
I've noticed a particular thing about the rarity of diamonds and mese, they don't seem to be as rare as I think they should be, in MC you have to dig for a long time before you get full diamond armor, whereas in MT you can get diamond armor in about an hour or two.

**Comment by Fixer-007** [comment]
While I'm playing minecraft beta, I can confirm - diamonds are extremely rare, iron is hard to come by, minetest game compared to that is bonanza :)

**Comment by ThomasMonroe314** [comment]
Come  to think of it, after a while it gets kind of boring, diamonds lose their worth, as does mese, etc

**Comment by ThomasMonroe314** [comment]
Meh, IMO it is still a little bit too high, I was thinking something like this:
Coal 0
Iron -64
Tin -128
Copper -256

Gold -1024
Diamond -4096
Mese crystal -2048
Mese block -4096

just to spread out the valuables and give incentive to actually mine at deeper and deeper levels

**Comment by paramat** [comment]
Certainly i would prefer that but we have to not break mods and games that rely on the old values, so i didn't change them radically.

For the 2nd density increase i'm thinking -8000.

**Comment by ThomasMonroe314** [comment]
> Certainly i would prefer that but we have to not break mods and games that rely on the old values, so i didn't change them radically.

@paramat I hate to disagree with you, but many of the subgames don't even use the up-to-date MTG at all or even the up-to-date "default", and I don't know of any mods that rely on the depth of the ores. (but you may know of a few :neutral_face: ) 

> For the 2nd density increase i'm thinking -8000. 

sounds good to me

**Comment by paramat** [comment]
Mods and games have been balanced with the current depths of ores, a dramatic change could be a problem. This is an example of how MTG is restricted by it being a mod base and so much relying on it. New subgames are needed that can be more free.

In a way the ease of reaching ores has not changed much because the 3D noise tunnels make it easier to get to a certain depth, you can sometimes walk to y = -256 without digging a node.

**Comment by ThomasMonroe314** [comment]
> Mods and games have been balanced with the current depths of ores, a dramatic change could be a problem. This is an example of how MTG is restricted by it being a mod base and so much relying on it. New subgames are needed that can be more free.

good point, but again, usually they have their own version of MTG that hasn't been updated, Legend Of Minetest for example.

> In a way the ease of reaching ores has not changed much because the 3D noise tunnels make it easier to get to a certain depth, you can sometimes walk to y = -256 without digging a node.

yeah that does make a difference, but it really depends on where you start mining.

**Comment by C1ffisme** [comment]
@paramat 
> Trying to decide between -8000 and -16000.

Okay... Lemme stop you there....

I don't think I'm going to even bother mining that far... Ever. Now, if caves got bigger and more common the further I went down, and there were underground biomes to keep the grey stone texture from becoming stale, I might be interested enough to go to -4000.

Heck, if you're gonna move ores this far down, a few underground "dimensions" like the Nether might be necessary to make sure the player doesn't get bored when mining thousands of blocks into the ground.

(Yes, I'm normally against dimension layers, but I'd gladly take them if it means I can see something that isn't lava, stone or ores while mining.)

**Comment by paramat** [comment]
I'll probably choose -8000, i know some have expressed there isn't much reason to go that deep, but the ores will be a reason, and we are assuming that a more interesting underground will be created later (which is the case). The 2nd increase has to be chosen on how MTG will be, chosen on the assumption of an interesting underground, otherwise we would have to change the depth later and end up with ores at an unsuitable depth.

The density increase at depth doesn't cause any harm even if done before underground is made more interesting, it adds a reward for those who go deep, and remember there are mods in use that add underground interest.

**Comment by ThomasMonroe314** [comment]
> I'm considering keeping the same number of ore clusters but doubling the ore number in each.

:+1:

**Comment by paramat** [comment]
Related PR #2047

**Comment by paramat** [comment]
I'm now unsure about increasing ore density at a deep level, it also conflicts with a much more interesting idea detailed in #2048 so i'm putting my former intention on hold for now.
Making some ores limited to certain stone types seems a more interesting direction.

**Comment by paramat** [comment]
With some discussion i agree with Ezhh that more ores lower down is not needed, it's better to make the exisitng lower ore regions deeper, which is what #2047 does.

Removing the overlap frees up resources for adding a new ore if that is ever done.

**Comment by Fixer-007** [comment]
So mining deep becomes totally pointless, unless there will be some nice deep biomes or something with dedicated stuffs.

**Comment by paramat** [comment]
There will be much more interesting things than high ore densities to travel down for, improving the underground is obviously already intended and is being discussed.

**Comment by scottwolff** [comment]
Imo, I think it is more like punishing the player. From a server admin side, it doesn't matter.
Anyways, I would assume ore spawns fall under game balance (challenge vs. reward) so falls under this argument and all of its fallacies and benefits: http://tvtropes.org/pmwiki/pmwiki.php/Main/DiminishingReturnsForBalance

**Comment by paramat** [comment]
> I think it is more like punishing the player

What do you refer to?

**Comment by Fixer-007** [comment]
Deeper ores?

**Comment by scottwolff** [comment]
Honrstly, i've played minecraft and the default design for orespawn is not so deep for some ores like gold, but it assumes you have bedrock and nether at lower depths. Also for some player obstacles like creepers that will blow apart crappier designs.
But i find that just having to dig to -256 rather than to -150 to be irritating.
Anyways, i can just change

But it does nag me like it is a design concept flaw.

But let's take a step back, procedural design is from the days of limited memory. 
Ergo, say if i was designing this for my server:

Deeper ores, is so that players don't make protection or more powerful and rarer or complicated stuff without actually taking more time on the server. Thus they don't take more memory or server resources from the server unless they become a dedicated loyal player.

Still players get better and better. Each player has their own playstyle. But dedicated and experienced players overcome the deeper ores whether you move them higher or down to start as low as -15000

It is like letting them get cut on pokemon later. What kind of dedign decision is that and does it really make it more challenging or out of balance? Does it matter changing the ore depths to such extremes? 
Then soon the draw is more pokemon types, more uniquely designed pokemon. You know, like more different types ores, more nodes. Im digressing.

**Comment by scottwolff** [comment]
*honestly

**Comment by scottwolff** [comment]
No I put an example, if we are talking about ore depth = difficulty. http://tvtropes.org/pmwiki/pmwiki.php/Main/DynamicDifficulty 
See, crash bandicoot, the designs were that if you kept getting destroyed by the boulder the game took this into account and made it less difficult. But for those players that were good enough to avoid the boulder it did not scale down the difficulty thus making the game easier where it needs to be but essentially the same game for all types of players.

**Comment by paramat** [comment]
#2107 merged. Keeping open for ores in particular stone types.

**Comment by paramat** [comment]
Closing due to #2710
